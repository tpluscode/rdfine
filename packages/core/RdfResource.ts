/* eslint-disable camelcase,@typescript-eslint/camelcase,no-dupe-class-members */
import RDF from '@rdfjs/data-model'
import { NamespaceBuilder } from '@rdfjs/namespace'
import { NamedNode, DatasetCore, BlankNode, DefaultGraph, Quad_Graph, Term, Literal } from 'rdf-js'
import cf, { SafeClownface, SingleContextClownface } from 'clownface'
import ResourceFactoryImpl, {
  Constructor,
  Mixin,
  ResourceCreationOptions,
  ResourceFactory,
  ResourceIndexer,
} from './lib/ResourceFactory'
import once from 'once'
import TypeCollectionCtor, { TypeCollection } from './lib/TypeCollection'
import { xsd } from '@tpluscode/rdf-ns-builders'

type ObjectOrFactory<T> = T | ((self: RdfResource) => T)

export type ResourceIdentifier = BlankNode | NamedNode
export type ResourceNode<D extends DatasetCore = DatasetCore> = { dataset: D; term: Term; graph?: NamedNode | DefaultGraph }

export interface GetOptions {
  strict: boolean
}

export interface RdfResource<D extends DatasetCore = DatasetCore> {
  readonly id: ResourceIdentifier
  readonly types: TypeCollection<D>
  readonly _selfGraph: SingleContextClownface<ResourceIdentifier, D>
  readonly _unionGraph: SafeClownface<ResourceIdentifier, D>
  readonly _graphId: Quad_Graph
  readonly _parent?: RdfResource<D>
  readonly isAnonymous: boolean
  hasType (type: string | NamedNode): boolean
  /**
   * Gets the value of a property
   * @param property
   */
  get<T extends RdfResource = RdfResource> (property: string | NamedNode): T
  get<T extends RdfResource = RdfResource> (property: string | NamedNode, options?: GetOptions): T | null
  /**
   * Gets the value of a property and ensures that an array will be returned
   * @param property
   */
  getArray<T extends RdfResource = RdfResource> (property: string | NamedNode, options?: GetOptions): T[]
  /**
   * Gets the property value if it's boolean. Throws if it's not
   * @param property
   */
  getBoolean (property: string | NamedNode, options?: GetOptions): boolean
  /**
   * Gets the property value if it's a string. Throws if it's not
   * @param property
   */
  getString (property: string | NamedNode): string
  getString (property: string | NamedNode, options?: GetOptions): string | null
  /**
   * Gets the property value if it's a number. Throws if it's not
   * @param property
   */
  getNumber (property: string | NamedNode): number | null
  getNumber (property: string | NamedNode, options?: GetOptions): number | null
  _getObjects(property: string | NamedNode, options?: GetOptions): SafeClownface<Term, D>
  _create<T extends RdfResource>(term: ResourceNode<D>, mixins?: Mixin[] | [Constructor, ...Mixin[]], options?: ResourceCreationOptions<D>): T & ResourceIndexer
}

export default class RdfResourceImpl<D extends DatasetCore = DatasetCore> implements RdfResource<D> {
  public readonly _selfGraph: SingleContextClownface<ResourceIdentifier, D>
  public readonly _unionGraph: SafeClownface<ResourceIdentifier, D>
  public readonly __initialized: boolean = false
  public readonly _parent?: RdfResource<D>
  private readonly __initializeProperties: (() => boolean)
  public static __ns?: NamespaceBuilder
  public static factory: ResourceFactory = new ResourceFactoryImpl(RdfResourceImpl)
  public static __mixins: Mixin[] = []

  private static _userInitializeProperties(resource: RdfResourceImpl, init: Initializer<RdfResource> = {}): void {
    Object.entries(init)
      .filter(([prop]) => prop !== 'id' && prop !== 'types')
      .forEach(([prop, value]) => {
        (resource as any)[prop] = value
      })

    if (init.types && Array.isArray(init.types)) {
      init.types.forEach(type => resource.types.add(type))
    }
  }

  public constructor(graph: ResourceNode<D>, init: Initializer<any> = {}, parent?: RdfResource<D>) {
    if (graph.term.termType !== 'BlankNode' && graph.term.termType !== 'NamedNode') {
      throw new Error(`RdfResource cannot be initialized from a ${graph.term.termType} node`)
    }

    const selfGraph = cf({
      ...graph,
      term: graph.term,
    })

    if (selfGraph._context[0].graph) {
      this._selfGraph = selfGraph
      this._unionGraph = cf({ dataset: selfGraph.dataset, term: selfGraph.term, graph: undefined })
    } else {
      this._selfGraph = cf({ dataset: selfGraph.dataset, term: selfGraph.term, graph: RDF.defaultGraph() })
      this._unionGraph = cf({ dataset: selfGraph.dataset, term: selfGraph.term })
    }

    this.__initializeProperties = once(() => {
      const self = this as any
      const defaults: Map<string, ObjectOrFactory<any>> = self.constructor.__defaults || new Map<string, ObjectOrFactory<any>>()
      defaults.forEach((value, key) => {
        const currentValue = self[key]
        const valueIsEmptyArray = Array.isArray(currentValue) && currentValue.length === 0
        const valueIsUndefined = typeof currentValue === 'undefined'

        if (!(valueIsEmptyArray || valueIsUndefined)) {
          return
        }

        if (typeof value === 'function') {
          self[key] = value(self)
          return
        }

        self[key] = value
      })

      return true
    })

    this._parent = parent
    this.__initialized = this.__initializeProperties()
    RdfResourceImpl._userInitializeProperties(this, init)
  }

  public get id(): ResourceIdentifier {
    return this._selfGraph.term
  }

  public get _graphId(): Quad_Graph {
    return this._selfGraph._context[0].graph!
  }

  public get types(): TypeCollection<D> {
    return new TypeCollectionCtor(this)
  }

  public get isAnonymous() {
    return this.id.termType === 'BlankNode'
  }

  public hasType(type: string | NamedNode): boolean {
    return this.types.has(type)
  }

  public equals(other: RdfResource | undefined | null): boolean {
    if (!other) {
      return false
    }

    return this.id.equals(other.id)
  }

  public get<T extends RdfResource = RdfResource>(property: string | NamedNode, options?: GetOptions): T | RdfResource<D> | null {
    const objects = this.getArray<T>(property, options)

    if (objects.length > 0) {
      return objects[0]
    }

    return null
  }

  public getArray<T extends RdfResource = RdfResource>(property: string | NamedNode, options?: GetOptions): T[] {
    const values = this._getObjects(property, options)
      .filter(obj => obj.term.termType === 'NamedNode' || obj.term.termType === 'BlankNode')
      .map(obj => {
        return this._create<T>(obj, [], { parent: this })
      })

    if (!values.length) {
      return []
    }

    return values
  }

  public getNumber(property: string | NamedNode): number
  public getNumber(property: string | NamedNode, options?: GetOptions): number | null {
    const [value] = this._getObjects(property, options).toArray()

    if (typeof value === 'undefined') {
      return null
    }

    if (value.term.termType === 'Literal') {
      return parseFloat(value.value)
    }

    throw new Error(`Expected property '${property}' to be a number but found '${value}'`)
  }

  public getString(property: string | NamedNode): string
  public getString(property: string | NamedNode, options?: GetOptions): string | null {
    const [value] = this._getObjects(property, options).toArray()

    if (typeof value === 'undefined') {
      return null
    }

    if (value.term.termType === 'Literal') {
      return value.value
    }

    throw new Error(`Expected property '${property}' to be a literal but found '${value}'`)
  }

  public getBoolean(property: string | NamedNode, options?: GetOptions): boolean {
    const [value] = this._getObjects(property, options).toArray()

    if (typeof value === 'undefined') {
      return false
    }

    if (value.term.termType === 'Literal' && xsd.boolean.equals(value.term.datatype)) {
      return value.term.equals(this._selfGraph.literal(true).term)
    }

    throw new Error(`Expected property '${property}' to be a boolean but found '${value}'`)
  }

  public _getObjects(property: string | NamedNode, { strict }: GetOptions = { strict: false }): SafeClownface<Term, D> {
    const propertyNode = typeof property === 'string' ? this._selfGraph.namedNode(property) : property
    const objects = this._selfGraph.out(propertyNode)

    if (!objects.terms.length && strict) {
      throw new Error(`Value for predicate <${property}> was missing`)
    }

    return objects
  }

  public _create<T extends RdfResource>(term: ResourceNode<D>, mixins?: Mixin[] | [Constructor, ...Mixin[]], options: ResourceCreationOptions<D> = {}): T & ResourceIndexer {
    return (this.constructor as Constructor).factory.createEntity<T>(term, mixins, options)
  }
}

type UserDefinedInterface<T extends RdfResource> = Omit<T, keyof RdfResource>

interface BaseInitializer {
  types?: NamedNode[] | TypeCollection<any>
  id?: RdfResource['id']
}

type InitialNode<Node extends Term = NamedNode | BlankNode> = Node | SingleContextClownface<Node>
type InitialLiteral = InitialNode<Literal>

export type Initializer<T> = Partial<Omit<{
  [P in keyof T]?:
  T[P] extends (infer U)[] ? U extends RdfResource ? Initializer<UserDefinedInterface<U>>[] | InitialNode[]
    : U extends Term ? T[P] | InitialNode<Term>[] : T[P] | InitialLiteral[]
    : T[P] extends RdfResource ? Initializer<UserDefinedInterface<T[P]> & BaseInitializer> | InitialNode :
      T[P] extends Term ? T[P] | InitialNode<T[P]>
        : T[P] | InitialLiteral
}, keyof RdfResource>> & BaseInitializer

type PartialRecursive<T> = T extends object ? { [K in keyof T]?: PartialRecursive<T[K]> } : T

export function fromObject<T extends RdfResource>(initializer: Initializer<UserDefinedInterface<T> & BaseInitializer>): T {
  return initializer as unknown as T
}
