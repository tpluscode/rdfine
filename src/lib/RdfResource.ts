import { Term, NamedNode, DatasetCore, BlankNode } from 'rdf-js'
import cf, { SingleContextClownface } from 'clownface'
import ns from '@rdfjs/namespace'
import once from 'once'

const rdf = ns('http://www.w3.org/1999/02/22-rdf-syntax-ns#')

export interface RdfResource<D extends DatasetCore = DatasetCore> {
  readonly id: Term | null
  readonly types: Term[]
  readonly _node: SingleContextClownface<D>
  hasType (type: string | NamedNode): boolean
}

export default class RdfResourceImpl<D extends DatasetCore = DatasetCore> implements RdfResource<D> {
  public readonly _node: SingleContextClownface<D>
  private readonly __initializeProperties: (() => void)
  public static __ns?: any
  private static __defaults = new Map<string, unknown>()

  public constructor(node: SingleContextClownface<D> | { dataset: D; term: NamedNode | BlankNode; graph?: NamedNode }) {
    const contexts = cf(node).toArray()

    if (contexts.length !== 1) {
      throw new Error('RdfResource can only be initialized from a single node. Got ' + contexts.length)
    }

    this._node = cf(node).toArray()[0]

    this.__initializeProperties = once(() => {
      const self = this as any
      RdfResourceImpl.__defaults.forEach((value, key) => {
        if (typeof self[key] === 'undefined') {
          self[key] = value
        }
      })
    })

    this.__initializeProperties()
  }

  public get id() {
    if (!this._node.term) throw new Error('Graph context does not represent a single node')

    return this._node.term
  }

  public get types() {
    return this._node.out(rdf.type).terms
  }

  public hasType(type: string | NamedNode) {
    const typeNode = typeof type === 'string' ? this._node.namedNode(type) : type

    return this._node.has(rdf.type, typeNode).terms.length > 0
  }
}
