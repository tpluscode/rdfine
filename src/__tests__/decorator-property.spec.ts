import { prefixes } from '@zazuko/rdf-vocabularies'
import cf from 'clownface'
import { namespace, property } from '..'
import RdfResource from '../lib/RdfResource'
import { parse, vocabs } from './_helpers'
import { Literal, NamedNode, Term } from 'rdf-js'
import { literal, namedNode } from 'rdf-data-model'
import rdfExt from 'rdf-ext'

const { ex, foaf, schema, rdf } = vocabs

describe('decorator', () => {
  describe('term', () => {
    describe('getter', () => {
      it('returns raw named node from property', async () => {
        // given
        const dataset = await parse(`
          @prefix ex: <${prefixes.ex}> .
          @prefix foaf: <${prefixes.foaf}> .

          ex:res foaf:friend ex:friend .`)
        class Resource extends RdfResource {
          @property({ path: foaf.friend })
          friend?: NamedNode
        }

        // when
        const instance = new Resource({
          dataset,
          term: ex.res,
        })
        const friend = instance.friend

        // then
        expect(friend!.value).toEqual(ex.friend.value)
      })

      it('returns raw blank node from property', async () => {
        // given
        const dataset = await parse(`
          @prefix ex: <${prefixes.ex}> .
          @prefix foaf: <${prefixes.foaf}> .
          @prefix schema: <${prefixes.schema}> .
          
          ex:res ex:likes [
            a ex:Blank
          ] .
        `)
        class Resource extends RdfResource {
          @property({ path: ex.likes })
          likes?: Term
        }

        // when
        const instance = new Resource({
          dataset,
          term: ex.res,
        })

        // then
        expect(instance.likes!.termType).toEqual('BlankNode')
      })

      it('returns raw literal node from property', async () => {
        // given
        const dataset = await parse(`
          @prefix ex: <${prefixes.ex}> .
          @prefix foaf: <${prefixes.foaf}> .
          @prefix schema: <${prefixes.schema}> .
          
          ex:res foaf:name "John Doe"@en-us .
        `)
        class Resource extends RdfResource {
          @property({ path: foaf.name })
          name?: Literal
        }

        // when
        const instance = new Resource({
          dataset,
          term: ex.res,
        })
        const name = instance.name

        // then
        expect(name!.value).toEqual('John Doe')
        expect(name!.language).toEqual('en-us')
      })

      it('throws for multiple values', async () => {
        // given
        const dataset = await parse(`
          @prefix ex: <${prefixes.ex}> .
          @prefix foaf: <${prefixes.foaf}> .
          @prefix schema: <${prefixes.schema}> .
          
          ex:res schema:children ex:Hansel, ex:Gretel .
        `)
        class Resource extends RdfResource {
          @property({ path: schema.children })
          children?: Literal
        }

        // when
        const instance = new Resource({
          dataset,
          term: ex.res,
        })

        // then
        expect(() => instance.children).toThrow()
      })

      it('return array when annotated', async () => {
        // given
        const dataset = await parse(`
          @prefix ex: <${prefixes.ex}> .
          @prefix schema: <${prefixes.schema}> .
          
          ex:res schema:children ex:Hansel, ex:Gretel .
        `)
        class Resource extends RdfResource {
          @property({ path: schema.children, array: true })
          children?: Term[]
        }

        // when
        const instance = new Resource({
          dataset,
          term: ex.res,
        })

        // then
        expect(instance.children).toHaveLength(2)
        expect(instance.children!.map(c => c.value)).toContain(ex.Hansel.value)
        expect(instance.children!.map(c => c.value)).toContain(ex.Gretel.value)
      })

      it('return correct node when annotated with namespace, using prop name', async () => {
        // given
        const dataset = await parse(`
          @prefix ex: <${prefixes.ex}> .
          @prefix foaf: <${prefixes.foaf}> .

          ex:res foaf:friend ex:friend .`)

        @namespace(foaf)
        class Resource extends RdfResource {
          @property()
          friend?: NamedNode
        }

        // when
        const instance = new Resource({
          dataset,
          term: ex.res,
        })
        const friend = instance.friend

        // then
        expect(friend!.value).toEqual(ex.friend.value)
      })

      it('get resource by paths of arbitrary length', async () => {
        // given
        const dataset = await parse(`
          @prefix ex: <${prefixes.ex}> .
          @prefix foaf: <${prefixes.foaf}> .

          ex:res foaf:friend ex:friend .
          ex:friend ex:worksAt ex:factory .
          ex:factory foaf:name "RDF Software House" .`)

        @namespace(foaf)
        class Resource extends RdfResource {
          @property({
            path: [
              'friend',
              'http://example.com/worksAt',
              foaf.name,
            ],
          })
          friendsWorkplaceName?: Literal
        }

        // when
        const instance = new Resource({
          dataset,
          term: ex.res,
        })

        // then
        expect(instance.friendsWorkplaceName!.value).toEqual('RDF Software House')
      })

      it('return correct node when annotated with namespace', async () => {
      // given
        const dataset = await parse(`
          @prefix ex: <${prefixes.ex}> .
          @prefix foaf: <${prefixes.foaf}> .

          ex:res foaf:friend ex:friend .`)

        @namespace(foaf)
        class Resource extends RdfResource {
          @property({ path: 'friend' })
          colleague?: NamedNode
        }

        // when
        const instance = new Resource({
          dataset,
          term: ex.res,
        })
        const friend = instance.colleague

        // then
        expect(friend!.value).toEqual(ex.friend.value)
      })

      it('strict throws when object not found', async () => {
        // given
        const dataset = rdfExt.dataset()
        class Resource extends RdfResource {
          @property({ path: foaf.name, strict: true })
          name!: Literal
        }

        // when
        const instance = new Resource({
          dataset,
          term: ex.res,
        })
        const getName = () => instance.name

        // then
        expect(getName).toThrow()
      })

      it('strict can be initialized', async () => {
        // given
        const dataset = rdfExt.dataset()
        class Resource extends RdfResource {
          @property.literal({ path: foaf.name, strict: true, initial: 'bar' })
          foo!: string
        }

        // when
        const instance = new Resource({
          dataset,
          term: ex.res,
        })

        // then
        expect(instance.foo).toEqual('bar')
        expect(dataset.toCanonical()).toMatchSnapshot()
      })

      it('returns rdf list array', async () => {
        // given
        const dataset = await parse(`
          @prefix ex: <${prefixes.ex}> .
          @prefix schema: <${prefixes.schema}> .
          
          ex:res ex:letters ( "a" "b" "c" ) .
        `)
        class Resource extends RdfResource {
          @property({ path: ex.letters, array: true })
          letters!: Literal[]
        }

        // when
        const instance = new Resource({
          dataset,
          term: ex.res,
        })

        // then
        expect(instance.letters.map(l => l.value)).toEqual(['a', 'b', 'c'])
      })

      it('returns empty rdf list array', async () => {
        // given
        const dataset = await parse(`
          @prefix ex: <${prefixes.ex}> .
          @prefix schema: <${prefixes.schema}> .
          
          ex:res ex:letters ( ) .
        `)
        class Resource extends RdfResource {
          @property({ path: ex.letters, array: true })
          letters!: Literal[]
        }

        // when
        const instance = new Resource({
          dataset,
          term: ex.res,
        })

        // then
        expect(instance.letters).toEqual([])
      })

      it('throws when rdf list is the object but not annotated', async () => {
        // given
        const dataset = await parse(`
          @prefix ex: <${prefixes.ex}> .
          @prefix foaf: <${prefixes.foaf}> .
          
          ex:res foaf:knows ( ex:Will ex:Joe ex:Sindy ) .
        `)
        class Resource extends RdfResource {
          @property({ path: foaf.knows })
          friend!: Term
        }

        // when
        const instance = new Resource({
          dataset,
          term: ex.res,
        })

        // then
        expect(() => instance.friend).toThrow()
      })
    })

    describe('setter', () => {
      it('can replace literal with blank', async () => {
        // given
        const dataset = await parse(`
          @prefix ex: <${prefixes.ex}> .
          @prefix foaf: <${prefixes.foaf}> .
          @prefix schema: <${prefixes.schema}> .
          
          ex:res foaf:name "John Doe"@en-us .
          [
            a ex:BlankNodeName ;
            ex:first "John" ;
            ex:last "Doe"
          ] .
        `)
        class Resource extends RdfResource {
          @property({ path: foaf.name })
          name?: Term | null
        }

        const instance = new Resource({
          dataset,
          term: ex.res,
        })

        // when
        instance.name = cf({ dataset }).has(rdf.type, ex.BlankNodeName).term

        // then
        expect(dataset.toCanonical()).toMatchSnapshot()
      })

      it('can set empty array, removing objects', async () => {
        // given
        const dataset = await parse(`
          @prefix ex: <${prefixes.ex}> .
          @prefix foaf: <${prefixes.foaf}> .
          @prefix schema: <${prefixes.schema}> .
          
          ex:res foaf:name "John"@en-us, "Johann"@de, "Jan"@pl, "Jean"@fr .
        `)
        class Resource extends RdfResource {
          @property({ path: foaf.name, array: true })
          name?: Term[]
        }

        const instance = new Resource({
          dataset,
          term: ex.res,
        })

        // when
        instance.name = []

        // then
        expect(dataset.toCanonical()).toMatchSnapshot()
      })

      it('can set array', async () => {
        // given
        const dataset = await parse(`
          @prefix ex: <${prefixes.ex}> .
          @prefix foaf: <${prefixes.foaf}> .
          @prefix schema: <${prefixes.schema}> .
          
          ex:res foaf:name "John"@en-us, "Johann"@de .
        `)
        class Resource extends RdfResource {
          @property({ path: foaf.name, array: true })
          name?: Term[]
        }

        const instance = new Resource({
          dataset,
          term: ex.res,
        })

        // when
        instance.name = [
          literal('Jan', 'pl'),
          literal('Jean', 'fr'),
        ]

        // then
        expect(dataset.toCanonical()).toMatchSnapshot()
      })

      it('throws when setting array to non-array getter', async () => {
        // given
        const dataset = await parse(`
          @prefix ex: <${prefixes.ex}> .
          @prefix foaf: <${prefixes.foaf}> .
          @prefix schema: <${prefixes.schema}> .
          
          ex:res foaf:name "John"@en-us, "Johann"@de, "Jan"@pl, "Jean"@fr .
        `)
        class Resource extends RdfResource {
          @property({ path: foaf.name })
          name?: any
        }

        const instance = new Resource({
          dataset,
          term: ex.res,
        })

        // when
        const setArray = () => { instance.name = [] }

        // then
        expect(setArray).toThrow()
      })

      it('sets value at paths of arbitrary length', async () => {
        // given
        const dataset = await parse(`
          @prefix ex: <${prefixes.ex}> .
          @prefix foaf: <${prefixes.foaf}> .

          ex:res foaf:friend ex:friend .
          ex:friend ex:worksAt ex:factory .
          ex:factory foaf:name "RDF Software House" .`)

        @namespace(foaf)
        class Resource extends RdfResource {
          @property({
            path: [
              'friend',
              'http://example.com/worksAt',
              foaf.name,
            ],
          })
          friendsWorkplaceName?: Literal
        }

        // when
        const instance = new Resource({
          dataset,
          term: ex.res,
        })
        instance.friendsWorkplaceName = literal('Google')

        // then
        expect(dataset.toCanonical()).toMatchSnapshot()
      })
    })

    describe('initial', () => {
      it('sets initial value from node', async () => {
        // given
        const dataset = rdfExt.dataset()
        class Resource extends RdfResource {
          @property({
            path: schema.name,
            initial: literal('foo'),
          })
          name!: Literal
        }

        // when
        const instance = new Resource({
          dataset,
          term: ex.res,
        })

        // then
        expect(instance.name.value).toEqual('foo')
        expect(dataset.toCanonical()).toMatchSnapshot()
      })

      it('sets initial value from function', async () => {
        // given
        const dataset = rdfExt.dataset()
        class Resource extends RdfResource {
          @property({
            path: schema.name,
            initial: (self: Resource) => namedNode(`${self.id.value}/child`),
          })
          child!: NamedNode
        }

        // when
        const instance = new Resource({
          dataset,
          term: ex.res,
        })

        // then
        expect(instance.child.value).toEqual('http://example.com/res/child')
        expect(dataset.toCanonical()).toMatchSnapshot()
      })

      it('throws when trying to replace rdf list', async () => {
        // given
        const dataset = await parse(`
          @prefix ex: <${prefixes.ex}> .
          @prefix foaf: <${prefixes.foaf}> .
          
          ex:res foaf:knows ( ex:Will ex:Joe ex:Sindy ) .
        `)
        class Resource extends RdfResource {
          @property({ path: foaf.knows, array: true })
          friend!: Term[]
        }

        // when
        const instance = new Resource({
          dataset,
          term: ex.res,
        })

        // then
        expect(() => {
          instance.friend = []
        }).toThrow()
      })
    })
  })
})
