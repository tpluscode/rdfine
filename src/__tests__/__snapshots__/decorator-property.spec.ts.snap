// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`decorator term getter array can be initialized 1`] = `
"<http://example.com/res> <http://xmlns.com/foaf/0.1/name> \\"bar\\" .
"
`;

exports[`decorator term getter array can be initialized with multiple values 1`] = `
"<http://example.com/res> <http://xmlns.com/foaf/0.1/name> \\"bar\\" .
<http://example.com/res> <http://xmlns.com/foaf/0.1/name> \\"foo\\" .
"
`;

exports[`decorator term getter rdf list can be initialized 1`] = `
"<http://example.com/res> <http://xmlns.com/foaf/0.1/name> _:c14n0 .
_:c14n0 <http://www.w3.org/1999/02/22-rdf-syntax-ns#first> \\"bar\\" .
_:c14n0 <http://www.w3.org/1999/02/22-rdf-syntax-ns#rest> <http://www.w3.org/1999/02/22-rdf-syntax-ns#nil> .
"
`;

exports[`decorator term getter rdf list can be initialized with multiple values 1`] = `
"<http://example.com/res> <http://xmlns.com/foaf/0.1/name> _:c14n0 .
_:c14n0 <http://www.w3.org/1999/02/22-rdf-syntax-ns#first> \\"foo\\" .
_:c14n0 <http://www.w3.org/1999/02/22-rdf-syntax-ns#rest> _:c14n1 .
_:c14n1 <http://www.w3.org/1999/02/22-rdf-syntax-ns#first> \\"bar\\" .
_:c14n1 <http://www.w3.org/1999/02/22-rdf-syntax-ns#rest> <http://www.w3.org/1999/02/22-rdf-syntax-ns#nil> .
"
`;

exports[`decorator term getter strict can be initialized 1`] = `
"<http://example.com/res> <http://xmlns.com/foaf/0.1/name> \\"bar\\" .
"
`;

exports[`decorator term initial sets initial value from function 1`] = `
"<http://example.com/res> <http://schema.org/name> <http://example.com/res/child> .
"
`;

exports[`decorator term initial sets initial value from node 1`] = `
"<http://example.com/res> <http://schema.org/name> \\"foo\\" .
"
`;

exports[`decorator term setter can replace literal with blank 1`] = `
"<http://example.com/res> <http://xmlns.com/foaf/0.1/name> _:c14n0 .
_:c14n0 <http://example.com/first> \\"John\\" .
_:c14n0 <http://example.com/last> \\"Doe\\" .
_:c14n0 <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://example.com/BlankNodeName> .
"
`;

exports[`decorator term setter can set array 1`] = `
"<http://example.com/res> <http://xmlns.com/foaf/0.1/name> \\"Jan\\"@pl .
<http://example.com/res> <http://xmlns.com/foaf/0.1/name> \\"Jean\\"@fr .
"
`;

exports[`decorator term setter can set empty array, removing objects 1`] = `""`;

exports[`decorator term setter sets nil for empty rdf list 1`] = `
"<http://example.com/res> <http://xmlns.com/foaf/0.1/knows> <http://www.w3.org/1999/02/22-rdf-syntax-ns#nil> .
"
`;

exports[`decorator term setter sets value at paths of arbitrary length 1`] = `
"<http://example.com/factory> <http://xmlns.com/foaf/0.1/name> \\"Google\\" .
<http://example.com/friend> <http://example.com/worksAt> <http://example.com/factory> .
<http://example.com/res> <http://xmlns.com/foaf/0.1/friend> <http://example.com/friend> .
"
`;

exports[`decorator term setter setting null to rdf list removes triple 1`] = `""`;
